File: parallel_dijkstra.py
#!/usr/bin/env python3
"""
并行Dijkstra最短路径算法实现
基于第三章的并行数值计算和并行图算法
"""

import time
import random
import heapq
from collections import defaultdict, deque
from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor
import multiprocessing as mp

class ParallelDijkstra:
    """并行Dijkstra最短路径算法类"""

    def __init__(self, strategy='sequential', num_threads=None):
        self.strategy = strategy
        self.num_threads = num_threads or mp.cpu_count()

    def sequential_dijkstra(self, graph, start_node):
        """串行Dijkstra算法实现"""
        # 初始化距离和前驱节点
        distances = {node: float('inf') for node in graph}
        distances[start_node] = 0
        previous = {node: None for node in graph}
        visited = set()

        # 优先队列：(distance, node)
        priority_queue = [(0, start_node)]

        while priority_queue:
            current_distance, current_node = heapq.heappop(priority_queue)

            # 如果已经访问过，跳过
            if current_node in visited:
                continue

            visited.add(current_node)

            # 更新邻居节点的距离
            for neighbor, weight in graph[current_node].items():
                if neighbor not in visited:
                    new_distance = current_distance + weight

                    if new_distance < distances[neighbor]:
                        distances[neighbor] = new_distance
                        previous[neighbor] = current_node
                        heapq.heappush(priority_queue, (new_distance, neighbor))

        return distances, previous

    def parallel_dijkstra(self, graph, start_node):
        """并行Dijkstra算法实现"""
        # 初始化
        distances = {node: float('inf') for node in graph}
        distances[start_node] = 0
        previous = {node: None for node in graph}
        visited = set()

        current_frontier = {start_node}

        while current_frontier:
            # 并行处理当前前沿的所有节点
            next_frontier = set()

            # 将前沿节点分组
            frontier_list = list(current_frontier)
            chunk_size = max(1, len(frontier_list) // self.num_threads)
            chunks = [frontier_list[i:i + chunk_size] for i in range(0, len(frontier_list), chunk_size)]

            with ThreadPoolExecutor(max_workers=self.num_threads) as executor:
                futures = []
                for chunk in chunks:
                    future = executor.submit(self._process_frontier_chunk, chunk, graph, distances, previous, visited)
                    futures.append(future)

                # 收集结果
                for future in futures:
                    chunk_next_frontier = future.result()
                    next_frontier.update(chunk_next_frontier)

            current_frontier = next_frontier

        return distances, previous

    def _process_frontier_chunk(self, chunk, graph, distances, previous, visited):
        """处理前沿块"""
        next_frontier = set()

        for current_node in chunk:
            if current_node in visited:
                continue

            visited.add(current_node)
            current_distance = distances[current_node]

            # 更新邻居节点的距离
            for neighbor, weight in graph[current_node].items():
                if neighbor not in visited:
                    new_distance = current_distance + weight

                    # 原子操作：使用锁保护距离更新
                    if new_distance < distances[neighbor]:
                        distances[neighbor] = new_distance
                        previous[neighbor] = current_node
                        next_frontier.add(neighbor)

        return next_frontier

    def bidirectional_dijkstra(self, graph, start, end):
        """双向Dijkstra算法"""
        if start == end:
            return 0, [start]

        # 正向搜索
        forward_distances = {node: float('inf') for node in graph}
        forward_distances[start] = 0
        forward_previous = {node: None for node in graph}
        forward_visited = set()
        forward_frontier = {start}

        # 反向搜索
        backward_distances = {node: float('inf') for node in graph}
        backward_distances[end] = 0
        backward_previous = {node: None for node in graph}
        backward_visited = set()
        backward_frontier = {end}

        meeting_node = None
        min_distance = float('inf')

        while forward_frontier and backward_frontier:
            # 并行处理正向前沿
            forward_next_frontier = set()
            forward_list = list(forward_frontier)
            chunk_size = max(1, len(forward_list) // self.num_threads)

            with ThreadPoolExecutor(max_workers=self.num_threads) as executor:
                futures = []
                for i in range(0, len(forward_list), chunk_size):
                    chunk = forward_list[i:i + chunk_size]
                    future = executor.submit(self._process_frontier_chunk_bidirectional,
                                           chunk, graph, forward_distances, forward_previous, forward_visited)
                    futures.append(future)

                for future in futures:
                    forward_next_frontier.update(future.result())

            forward_frontier = forward_next_frontier

            # 检查相遇点
            for node in forward_visited:
                if node in backward_visited:
                    total_distance = forward_distances[node] + backward_distances[node]
                    if total_distance < min_distance:
                        min_distance = total_distance
                        meeting_node = node

            if meeting_node:
                break

            # 并行处理反向前沿
            backward_next_frontier = set()
            backward_list = list(backward_frontier)
            chunk_size = max(1, len(backward_list) // self.num_threads)

            with ThreadPoolExecutor(max_workers=self.num_threads) as executor:
                futures = []
                for i in range(0, len(backward_list), chunk_size):
                    chunk = backward_list[i:i + chunk_size]
                    future = executor.submit(self._process_frontier_chunk_bidirectional_reverse,
                                           chunk, graph, backward_distances, backward_previous, backward_visited)
                    futures.append(future)

                for future in futures:
                    backward_next_frontier.update(future.result())

            backward_frontier = backward_next_frontier

            # 检查相遇点
            for node in backward_visited:
                if node in forward_visited:
                    total_distance = forward_distances[node] + backward_distances[node]
                    if total_distance < min_distance:
                        min_distance = total_distance
                        meeting_node = node

            if meeting_node:
                break

        if meeting_node is None:
            return float('inf'), []

        # 重构路径
        path = []
        current = meeting_node

        # 从meeting_node到start
        while current is not None:
            path.append(current)
            current = forward_previous[current]

        path.reverse()

        # 从meeting_node到end
        current = backward_previous[meeting_node]
        while current is not None:
            path.append(current)
            current = backward_previous[current]

        return min_distance, path

    def _process_frontier_chunk_bidirectional(self, chunk, graph, distances, previous, visited):
        """双向Dijkstra的正向处理"""
        next_frontier = set()

        for current_node in chunk:
            if current_node in visited:
                continue

            visited.add(current_node)
            current_distance = distances[current_node]

            for neighbor, weight in graph[current_node].items():
                if neighbor not in visited:
                    new_distance = current_distance + weight
                    if new_distance < distances[neighbor]:
                        distances[neighbor] = new_distance
                        previous[neighbor] = current_node
                        next_frontier.add(neighbor)

        return next_frontier

    def _process_frontier_chunk_bidirectional_reverse(self, chunk, graph, distances, previous, visited):
        """双向Dijkstra的反向处理"""
        next_frontier = set()

        for current_node in chunk:
            if current_node in visited:
                continue

            visited.add(current_node)
            current_distance = distances[current_node]

            for neighbor, weight in graph[current_node].items():
                if neighbor not in visited:
                    new_distance = current_distance + weight
                    if new_distance < distances[neighbor]:
                        distances[neighbor] = new_distance
                        previous[neighbor] = current_node
                        next_frontier.add(neighbor)

        return next_frontier

    def parallel_a_star(self, graph, start, end, heuristic_func):
        """并行A*算法"""
        # 初始化
        distances = {node: float('inf') for node in graph}
        distances[start] = 0
        previous = {node: None for node in graph}
        visited = set()

        # 优先队列：(f_score, g_score, node)
        priority_queue = [(heuristic_func(start, end), 0, start)]

        while priority_queue:
            f_score, current_distance, current_node = heapq.heappop(priority_queue)

            if current_node in visited:
                continue

            if current_node == end:
                break

            visited.add(current_node)

            # 并行处理邻居节点
            neighbors = list(graph[current_node].items())
            chunk_size = max(1, len(neighbors) // self.num_threads)
            chunks = [neighbors[i:i + chunk_size] for i in range(0, len(neighbors), chunk_size)]

            next_frontier = []

            with ThreadPoolExecutor(max_workers=self.num_threads) as executor:
                futures = []
                for chunk in chunks:
                    future = executor.submit(self._process_neighbors_chunk,
                                           chunk, current_node, current_distance, distances, previous, visited, heuristic_func, end)
                    futures.append(future)

                for future in futures:
                    next_frontier.extend(future.result())

            # 将新的前沿节点加入优先队列
            for neighbor, new_distance in next_frontier:
                f_score = new_distance + heuristic_func(neighbor, end)
                heapq.heappush(priority_queue, (f_score, new_distance, neighbor))

        return distances, previous

    def _process_neighbors_chunk(self, chunk, current_node, current_distance, distances, previous, visited, heuristic_func, end):
        """处理邻居节点块"""
        next_frontier = []

        for neighbor, weight in chunk:
            if neighbor not in visited:
                new_distance = current_distance + weight

                if new_distance < distances[neighbor]:
                    distances[neighbor] = new_distance
                    previous[neighbor] = current_node
                    next_frontier.append((neighbor, new_distance))

        return next_frontier

    def get_shortest_path(self, graph, start, end, method='sequential'):
        """获取最短路径"""
        if method == 'sequential':
            distances, previous = self.sequential_dijkstra(graph, start)
        elif method == 'parallel':
            distances, previous = self.parallel_dijkstra(graph, start)
        elif method == 'bidirectional':
            return self.bidirectional_dijkstra(graph, start, end)
        else:
            raise ValueError(f"Unsupported method: {method}")

        if distances[end] == float('inf'):
            return float('inf'), []

        # 重构路径
        path = []
        current = end
        while current is not None:
            path.append(current)
            current = previous[current]

        path.reverse()
        return distances[end], path

def generate_random_weighted_graph(n, edge_probability=0.1, max_weight=10):
    """生成随机带权图"""
    graph = defaultdict(dict)

    for i in range(n):
        for j in range(i + 1, n):
            if random.random() < edge_probability:
                weight = random.randint(1, max_weight)
                graph[i][j] = weight
                graph[j][i] = weight

    return graph

def manhattan_distance(node1, node2):
    """曼哈顿距离启发函数"""
    # 假设节点是坐标形式 (x, y)
    if isinstance(node1, int) and isinstance(node2, int):
        # 如果节点是整数，使用简单的差值
        return abs(node1 - node2)
    return abs(node1[0] - node2[0]) + abs(node1[1] - node2[1])

def benchmark_dijkstra_algorithms(graph_size=1000, num_tests=3):
    """Dijkstra算法性能基准测试"""
    print(f"=== Dijkstra算法性能测试 ===")
    print(f"图大小: {graph_size} 个节点")
    print(f"CPU核心数: {mp.cpu_count()}")
    print()

    # 生成测试图
    graph = generate_random_weighted_graph(graph_size, 0.05, 10)
    start_node = 0
    end_node = graph_size - 1

    strategies = [
        ('串行Dijkstra', 'sequential'),
        ('并行Dijkstra', 'parallel'),
        ('双向Dijkstra', 'bidirectional'),
    ]

    results = {}

    for name, method in strategies:
        print(f"测试: {name}")
        times = []

        for i in range(num_tests):
            start_time = time.time()
            dijkstra = ParallelDijkstra()
            distance, path = dijkstra.get_shortest_path(graph, start_node, end_node, method)
            end_time = time.time()

            execution_time = end_time - start_time
            times.append(execution_time)

        avg_time = sum(times) / len(times)
        min_time = min(times)
        results[method] = avg_time

        print(f"  平均时间: {avg_time:.4f}s")
        print(f"  最佳时间: {min_time:.4f}s")
        print(f"  最短距离: {distance}")
        print()

    # 计算加速比
    print("=== 加速比分析 ===")
    sequential_time = results['sequential']

    for name, method in strategies:
        if method != 'sequential':
            speedup = sequential_time / results[method]
            efficiency = speedup / mp.cpu_count()
            print(f"{name}:")
            print(f"  加速比: {speedup:.2f}x")
            print(f"  效率: {efficiency:.2%}")
            print()

def demo_parallel_dijkstra():
    """演示并行Dijkstra算法"""
    print("=== 并行Dijkstra算法演示 ===")

    # 小规模图演示
    graph = {
        0: {1: 4, 2: 2},
        1: {0: 4, 2: 1, 3: 5},
        2: {0: 2, 1: 1, 3: 8, 4: 10},
        3: {1: 5, 2: 8, 4: 2},
        4: {2: 10, 3: 2}
    }

    print("图结构:")
    for node, neighbors in graph.items():
        print(f"  {node}: {neighbors}")

    start_node = 0
    end_node = 4

    # 串行Dijkstra
    dijkstra = ParallelDijkstra(strategy='sequential')
    distance, path = dijkstra.get_shortest_path(graph, start_node, end_node, 'sequential')

    print(f"\n从节点 {start_node} 到 {end_node} 的最短路径:")
    print(f"  距离: {distance}")
    print(f"  路径: {' -> '.join(map(str, path))}")

    # 并行Dijkstra
    distance_parallel, path_parallel = dijkstra.get_shortest_path(graph, start_node, end_node, 'parallel')
    print(f"\n并行Dijkstra结果:")
    print(f"  距离: {distance_parallel}")
    print(f"  路径: {' -> '.join(map(str, path_parallel))}")

    # 双向Dijkstra
    distance_bidirectional, path_bidirectional = dijkstra.bidirectional_dijkstra(graph, start_node, end_node)
    print(f"\n双向Dijkstra结果:")
    print(f"  距离: {distance_bidirectional}")
    print(f"  路径: {' -> '.join(map(str, path_bidirectional))}")

    print()

    # 性能测试
    benchmark_dijkstra_algorithms()

if __name__ == "__main__":
    demo_parallel_dijkstra()

File: parallel_dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <thread>
#include <future>
#include <random>
#include <chrono>
#include <memory>
#include <functional>
#include <algorithm>
#include <climits>

class ParallelDijkstra {
private:
    int num_threads_;

    // 串行Dijkstra算法实现
    std::pair<std::unordered_map<int, int>, std::unordered_map<int, int>>
    sequential_dijkstra(const std::unordered_map<int, std::unordered_map<int, int>>& graph, int start_node) {
        std::unordered_map<int, int> distances;
        std::unordered_map<int, int> previous;
        std::unordered_set<int> visited;

        // 初始化
        for (const auto& [node, _] : graph) {
            distances[node] = INT_MAX;
            previous[node] = -1;
        }
        distances[start_node] = 0;

        // 优先队列：(distance, node)
        std::priority_queue<std::pair<int, int>, std::vector<std::pair<int, int>>, std::greater<std::pair<int, int>>> pq;
        pq.push({0, start_node});

        while (!pq.empty()) {
            auto [current_distance, current_node] = pq.top();
            pq.pop();

            // 如果已经访问过，跳过
            if (visited.find(current_node) != visited.end()) {
                continue;
            }

            visited.insert(current_node);

            // 更新邻居节点的距离
            for (const auto& [neighbor, weight] : graph.at(current_node)) {
                if (visited.find(neighbor) == visited.end()) {
                    int new_distance = current_distance + weight;

                    if (new_distance < distances[neighbor]) {
                        distances[neighbor] = new_distance;
                        previous[neighbor] = current_node;
                        pq.push({new_distance, neighbor});
                    }
                }
            }
        }

        return {distances, previous};
    }

    // 并行Dijkstra算法实现
    std::pair<std::unordered_map<int, int>, std::unordered_map<int, int>>
    parallel_dijkstra(const std::unordered_map<int, std::unordered_map<int, int>>& graph, int start_node) {
        std::unordered_map<int, int> distances;
        std::unordered_map<int, int> previous;
        std::unordered_set<int> visited;

        // 初始化
        for (const auto& [node, _] : graph) {
            distances[node] = INT_MAX;
            previous[node] = -1;
        }
        distances[start_node] = 0;

        std::unordered_set<int> current_frontier;
        current_frontier.insert(start_node);

        while (!current_frontier.empty()) {
            std::unordered_set<int> next_frontier;

            // 将前沿节点分组
            std::vector<int> frontier_list(current_frontier.begin(), current_frontier.end());
            int chunk_size = std::max(1, static_cast<int>(frontier_list.size()) / num_threads_);

            std::vector<std::future<std::vector<int>>> futures;

            for (int i = 0; i < frontier_list.size(); i += chunk_size) {
                int end = std::min(i + chunk_size, static_cast<int>(frontier_list.size()));
                std::vector<int> chunk(frontier_list.begin() + i, frontier_list.begin() + end);

                auto future = std::async(std::launch::async, [&, chunk]() {
                    std::vector<int> chunk_next_frontier;
                    std::mutex mutex;

                    for (int current_node : chunk) {
                        if (visited.find(current_node) != visited.end()) {
                            continue;
                        }

                        std::lock_guard<std::mutex> lock(mutex);
                        if (visited.find(current_node) != visited.end()) {
                            continue;
                        }

                        visited.insert(current_node);
                        int current_distance = distances[current_node];

                        // 更新邻居节点的距离
                        for (const auto& [neighbor, weight] : graph.at(current_node)) {
                            if (visited.find(neighbor) == visited.end()) {
                                int new_distance = current_distance + weight;

                                if (new_distance < distances[neighbor]) {
                                    distances[neighbor] = new_distance;
                                    previous[neighbor] = current_node;
                                    chunk_next_frontier.push_back(neighbor);
                                }
                            }
                        }
                    }

                    return chunk_next_frontier;
                });

                futures.push_back(std::move(future));
            }

            // 收集结果
            for (auto& future : futures) {
                auto chunk_next_frontier = future.get();
                next_frontier.insert(chunk_next_frontier.begin(), chunk_next_frontier.end());
            }

            current_frontier = std::move(next_frontier);
        }

        return {distances, previous};
    }

    // 双向Dijkstra算法
    std::pair<int, std::vector<int>>
    bidirectional_dijkstra(const std::unordered_map<int, std::unordered_map<int, int>>& graph, int start, int end) {
        if (start == end) {
            return {0, {start}};
        }

        // 正向搜索
        std::unordered_map<int, int> forward_distances;
        std::unordered_map<int, int> forward_previous;
        std::unordered_set<int> forward_visited;
        std::unordered_set<int> forward_frontier;

        // 反向搜索
        std::unordered_map<int, int> backward_distances;
        std::unordered_map<int, int> backward_previous;
        std::unordered_set<int> backward_visited;
        std::unordered_set<int> backward_frontier;

        // 初始化
        for (const auto& [node, _] : graph) {
            forward_distances[node] = INT_MAX;
            forward_previous[node] = -1;
            backward_distances[node] = INT_MAX;
            backward_previous[node] = -1;
        }

        forward_distances[start] = 0;
        backward_distances[end] = 0;
        forward_frontier.insert(start);
        backward_frontier.insert(end);

        int meeting_node = -1;
        int min_distance = INT_MAX;

        while (!forward_frontier.empty() && !backward_frontier.empty()) {
            // 并行处理正向前沿
            std::unordered_set<int> forward_next_frontier;
            std::vector<int> forward_list(forward_frontier.begin(), forward_frontier.end());
            int chunk_size = std::max(1, static_cast<int>(forward_list.size()) / num_threads_);

            std::vector<std::future<std::vector<int>>> forward_futures;

            for (int i = 0; i < forward_list.size(); i += chunk_size) {
                int end_idx = std::min(i + chunk_size, static_cast<int>(forward_list.size()));
                std::vector<int> chunk(forward_list.begin() + i, forward_list.begin() + end_idx);

                auto future = std::async(std::launch::async, [&, chunk]() {
                    std::vector<int> chunk_next_frontier;
                    std::mutex mutex;

                    for (int current_node : chunk) {
                        if (forward_visited.find(current_node) != forward_visited.end()) {
                            continue;
                        }

                        std::lock_guard<std::mutex> lock(mutex);
                        if (forward_visited.find(current_node) != forward_visited.end()) {
                            continue;
                        }

                        forward_visited.insert(current_node);
                        int current_distance = forward_distances[current_node];

                        for (const auto& [neighbor, weight] : graph.at(current_node)) {
                            if (forward_visited.find(neighbor) == forward_visited.end()) {
                                int new_distance = current_distance + weight;
                                if (new_distance < forward_distances[neighbor]) {
                                    forward_distances[neighbor] = new_distance;
                                    forward_previous[neighbor] = current_node;
                                    chunk_next_frontier.push_back(neighbor);
                                }
                            }
                        }
                    }

                    return chunk_next_frontier;
                });

                forward_futures.push_back(std::move(future));
            }

            for (auto& future : forward_futures) {
                auto chunk_next_frontier = future.get();
                forward_next_frontier.insert(chunk_next_frontier.begin(), chunk_next_frontier.end());
            }

            forward_frontier = std::move(forward_next_frontier);

            // 检查相遇点
            for (int node : forward_visited) {
                if (backward_visited.find(node) != backward_visited.end()) {
                    int total_distance = forward_distances[node] + backward_distances[node];
                    if (total_distance < min_distance) {
                        min_distance = total_distance;
                        meeting_node = node;
                    }
                }
            }

            if (meeting_node != -1) {
                break;
            }

            // 并行处理反向前沿
            std::unordered_set<int> backward_next_frontier;
            std::vector<int> backward_list(backward_frontier.begin(), backward_frontier.end());
            chunk_size = std::max(1, static_cast<int>(backward_list.size()) / num_threads_);

            std::vector<std::future<std::vector<int>>> backward_futures;

            for (int i = 0; i < backward_list.size(); i += chunk_size) {
                int end_idx = std::min(i + chunk_size, static_cast<int>(backward_list.size()));
                std::vector<int> chunk(backward_list.begin() + i, backward_list.begin() + end_idx);

                auto future = std::async(std::launch::async, [&, chunk]() {
                    std::vector<int> chunk_next_frontier;
                    std::mutex mutex;

                    for (int current_node : chunk) {
                        if (backward_visited.find(current_node) != backward_visited.end()) {
                            continue;
                        }

                        std::lock_guard<std::mutex> lock(mutex);
                        if (backward_visited.find(current_node) != backward_visited.end()) {
                            continue;
                        }

                        backward_visited.insert(current_node);
                        int current_distance = backward_distances[current_node];

                        for (const auto& [neighbor, weight] : graph.at(current_node)) {
                            if (backward_visited.find(neighbor) == backward_visited.end()) {
                                int new_distance = current_distance + weight;
                                if (new_distance < backward_distances[neighbor]) {
                                    backward_distances[neighbor] = new_distance;
                                    backward_previous[neighbor] = current_node;
                                    chunk_next_frontier.push_back(neighbor);
                                }
                            }
                        }
                    }

                    return chunk_next_frontier;
                });

                backward_futures.push_back(std::move(future));
            }

            for (auto& future : backward_futures) {
                auto chunk_next_frontier = future.get();
                backward_next_frontier.insert(chunk_next_frontier.begin(), chunk_next_frontier.end());
            }

            backward_frontier = std::move(backward_next_frontier);

            // 检查相遇点
            for (int node : backward_visited) {
                if (forward_visited.find(node) != forward_visited.end()) {
                    int total_distance = forward_distances[node] + backward_distances[node];
                    if (total_distance < min_distance) {
                        min_distance = total_distance;
                        meeting_node = node;
                    }
                }
            }

            if (meeting_node != -1) {
                break;
            }
        }

        if (meeting_node == -1) {
            return {INT_MAX, {}};
        }

        // 重构路径
        std::vector<int> path;
        int current = meeting_node;

        // 从meeting_node到start
        while (current != -1) {
            path.push_back(current);
            current = forward_previous[current];
        }

        std::reverse(path.begin(), path.end());

        // 从meeting_node到end
        current = backward_previous[meeting_node];
        while (current != -1) {
            path.push_back(current);
            current = backward_previous[current];
        }

        return {min_distance, path};
    }

public:
    ParallelDijkstra(int num_threads = -1) {
        num_threads_ = (num_threads == -1) ? std::thread::hardware_concurrency() : num_threads;
    }

    // 主要的Dijkstra函数
    std::pair<std::unordered_map<int, int>, std::unordered_map<int, int>>
    dijkstra(const std::unordered_map<int, std::unordered_map<int, int>>& graph, int start_node, const std::string& method = "sequential") {
        if (method == "sequential") {
            return sequential_dijkstra(graph, start_node);
        } else if (method == "parallel") {
            return parallel_dijkstra(graph, start_node);
        } else {
            throw std::invalid_argument("不支持的方法: " + method);
        }
    }

    // 获取最短路径
    std::pair<int, std::vector<int>>
    get_shortest_path(const std::unordered_map<int, std::unordered_map<int, int>>& graph, int start, int end, const std::string& method = "sequential") {
        if (method == "sequential") {
            auto [distances, previous] = sequential_dijkstra(graph, start);
            if (distances[end] == INT_MAX) {
                return {INT_MAX, {}};
            }

            std::vector<int> path;
            int current = end;
            while (current != -1) {
                path.push_back(current);
                current = previous[current];
            }

            std::reverse(path.begin(), path.end());
            return {distances[end], path};

        } else if (method == "parallel") {
            auto [distances, previous] = parallel_dijkstra(graph, start);
            if (distances[end] == INT_MAX) {
                return {INT_MAX, {}};
            }

            std::vector<int> path;
            int current = end;
            while (current != -1) {
                path.push_back(current);
                current = previous[current];
            }

            std::reverse(path.begin(), path.end());
            return {distances[end], path};

        } else if (method == "bidirectional") {
            return bidirectional_dijkstra(graph, start, end);

        } else {
            throw std::invalid_argument("不支持的方法: " + method);
        }
    }
};

// 生成随机带权图
std::unordered_map<int, std::unordered_map<int, int>>
generate_random_weighted_graph(int n, double edge_probability = 0.05, int max_weight = 10) {
    std::unordered_map<int, std::unordered_map<int, int>> graph;
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(0.0, 1.0);
    std::uniform_int_distribution<> weight_dis(1, max_weight);

    for (int i = 0; i < n; i++) {
        graph[i] = std::unordered_map<int, int>();
    }

    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            if (dis(gen) < edge_probability) {
                int weight = weight_dis(gen);
                graph[i][j] = weight;
                graph[j][i] = weight;
            }
        }
    }

    return graph;
}

// 性能测试工具
class PerformanceTimer {
private:
    std::chrono::high_resolution_clock::time_point start_time;

public:
    void start() {
        start_time = std::chrono::high_resolution_clock::now();
    }

    double elapsed_seconds() {
        auto end_time = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end_time - start_time);
        return duration.count() / 1000000.0;
    }
};

// 性能基准测试
void benchmark_dijkstra_algorithms(int graph_size = 1000, int num_tests = 3) {
    std::cout << "=== Dijkstra算法性能测试 ===" << std::endl;
    std::cout << "图大小: " << graph_size << " 个节点" << std::endl;
    std::cout << "CPU核心数: " << std::thread::hardware_concurrency() << std::endl;
    std::cout << std::endl;

    // 生成测试图
    auto graph = generate_random_weighted_graph(graph_size, 0.05, 10);
    int start_node = 0;
    int end_node = graph_size - 1;

    std::vector<std::pair<std::string, std::string>> strategies = {
        {"串行Dijkstra", "sequential"},
        {"并行Dijkstra", "parallel"},
        {"双向Dijkstra", "bidirectional"}
    };

    std::vector<double> results;

    for (const auto& strategy : strategies) {
        std::cout << "测试: " << strategy.first << std::endl;

        std::vector<double> times;
        for (int i = 0; i < num_tests; i++) {
            PerformanceTimer timer;
            timer.start();

            ParallelDijkstra dijkstra;
            auto [distance, path] = dijkstra.get_shortest_path(graph, start_node, end_node, strategy.second);

            double execution_time = timer.elapsed_seconds();
            times.push_back(execution_time);
        }

        double avg_time = 0.0;
        for (double time : times) avg_time += time;
        avg_time /= times.size();

        double min_time = *std::min_element(times.begin(), times.end());
        results.push_back(avg_time);

        std::cout << "  平均时间: " << avg_time << "s" << std::endl;
        std::cout << "  最佳时间: " << min_time << "s" << std::endl;
        std::cout << std::endl;
    }

    // 计算加速比
    std::cout << "=== 加速比分析 ===" << std::endl;
    double sequential_time = results[0];

    for (size_t i = 1; i < results.size(); i++) {
        double speedup = sequential_time / results[i];
        double efficiency = speedup / std::thread::hardware_concurrency();

        std::cout << strategies[i].first << ":" << std::endl;
        std::cout << "  加速比: " << speedup << "x" << std::endl;
        std::cout << "  效率: " << (efficiency * 100) << "%" << std::endl;
        std::cout << std::endl;
    }
}

// 演示函数
void demo_parallel_dijkstra() {
    std::cout << "=== 并行Dijkstra算法演示 ===" << std::endl;

    // 小规模图演示
    std::unordered_map<int, std::unordered_map<int, int>> graph = {
        {0, {{1, 4}, {2, 2}}},
        {1, {{0, 4}, {2, 1}, {3, 5}}},
        {2, {{0, 2}, {1, 1}, {3, 8}, {4, 10}}},
        {3, {{1, 5}, {2, 8}, {4, 2}}},
        {4, {{2, 10}, {3, 2}}}
    };

    std::cout << "图结构:" << std::endl;
    for (const auto& [node, neighbors] : graph) {
        std::cout << "  " << node << ": ";
        for (const auto& [neighbor, weight] : neighbors) {
            std::cout << neighbor << "(" << weight << ") ";
        }
        std::cout << std::endl;
    }

    int start_node = 0;
    int end_node = 4;

    // 串行Dijkstra
    ParallelDijkstra dijkstra;
    auto [distance, path] = dijkstra.get_shortest_path(graph, start_node, end_node, "sequential");

    std::cout << "\n从节点 " << start_node << " 到 " << end_node << " 的最短路径:" << std::endl;
    std::cout << "  距离: " << distance << std::endl;
    std::cout << "  路径: ";
    for (size_t i = 0; i < path.size(); i++) {
        std::cout << path[i];
        if (i < path.size() - 1) std::cout << " -> ";
    }
    std::cout << std::endl;

    // 并行Dijkstra
    auto [distance_parallel, path_parallel] = dijkstra.get_shortest_path(graph, start_node, end_node, "parallel");
    std::cout << "\n并行Dijkstra结果:" << std::endl;
    std::cout << "  距离: " << distance_parallel << std::endl;
    std::cout << "  路径: ";
    for (size_t i = 0; i < path_parallel.size(); i++) {
        std::cout << path_parallel[i];
        if (i < path_parallel.size() - 1) std::cout << " -> ";
    }
    std::cout << std::endl;

    // 双向Dijkstra
    auto [distance_bidirectional, path_bidirectional] = dijkstra.get_shortest_path(graph, start_node, end_node, "bidirectional");
    std::cout << "\n双向Dijkstra结果:" << std::endl;
    std::cout << "  距离: " << distance_bidirectional << std::endl;
    std::cout << "  路径: ";
    for (size_t i = 0; i < path_bidirectional.size(); i++) {
        std::cout << path_bidirectional[i];
        if (i < path_bidirectional.size() - 1) std::cout << " -> ";
    }
    std::cout << std::endl;

    std::cout << std::endl;

    // 性能测试
    benchmark_dijkstra_algorithms();
}

int main() {
    demo_parallel_dijkstra();
    return 0;
}